File: packages/sdk-android/src/main/java/com/lynxship/sdk/android/LynxShipGlobalProps.java
package com.lynxship.sdk.android;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Build;
import android.os.PowerManager;
import android.view.View;
import android.view.WindowInsets;
import android.view.accessibility.AccessibilityManager;

import androidx.annotation.NonNull;

import com.lynx.tasm.TemplateData;
import com.lynx.tasm.LynxView;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/** Builds the reserved host context injected into an Android Lynx container. */
final class LynxShipGlobalProps {
    private LynxShipGlobalProps() {}

    static void apply(
            @NonNull LynxView lynxView,
            @NonNull View host,
            @NonNull String containerId,
            @NonNull String containerInitTime,
            @NonNull Map<String, Object> appProps,
            boolean autoEnabled,
            boolean appInBackground) {
        if (!autoEnabled && appProps.isEmpty()) return;
        Map<String, Object> props = autoEnabled
                ? create(host, containerId, containerInitTime, appProps, appInBackground)
                : appProps;
        lynxView.updateGlobalProps(TemplateData.fromMap(props));
    }

    @NonNull
    static Map<String, Object> create(
            @NonNull View host,
            @NonNull String containerId,
            @NonNull String containerInitTime,
            @NonNull Map<String, Object> appProps,
            boolean appInBackground) {
        Context context = host.getContext();
        float density = context.getResources().getDisplayMetrics().density;
        if (!(density > 0f) || !Float.isFinite(density)) density = 1f;
        double screenWidth = context.getResources().getDisplayMetrics().widthPixels / (double) density;
        double screenHeight = context.getResources().getDisplayMetrics().heightPixels / (double) density;
        double contentWidth = Math.max(0d, host.getWidth() / (double) density);
        double contentHeight = Math.max(0d, host.getHeight() / (double) density);
        Map<String, Double> insets = systemInsets(host, density);
        String orientation = context.getResources().getConfiguration().orientation ==
                Configuration.ORIENTATION_LANDSCAPE ? "landscape" : "portrait";
        int night = context.getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;
        String theme = night == Configuration.UI_MODE_NIGHT_YES ? "dark" : "light";
        String locale = context.getResources().getConfiguration().getLocales().isEmpty()
                ? "en-US" : context.getResources().getConfiguration().getLocales().get(0).toLanguageTag();
        PowerManager power = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        boolean cutout = Build.VERSION.SDK_INT >= Build.VERSION_CODES.P &&
                host.getRootWindowInsets() != null && host.getRootWindowInsets().getDisplayCutout() != null;
        boolean tablet = context.getResources().getConfiguration().smallestScreenWidthDp >= 600;
        Map<String, Object> props = new HashMap<>(appProps);
        props.put("os", "android");
        props.put("osVersion", Build.VERSION.RELEASE == null ? "unknown" : Build.VERSION.RELEASE);
        props.put("deviceModel", Build.MODEL == null ? "unknown" : Build.MODEL);
        props.put("containerID", containerId);
        props.put("containerInitTime", containerInitTime);
        props.put("screenWidth", screenWidth);
        props.put("screenHeight", screenHeight);
        props.put("contentWidth", contentWidth);
        props.put("contentHeight", contentHeight);
        props.put("safeAreaInsets", insets);
        props.put("pixelRatio", (double) density);
        props.put("accessibleMode", accessibilityMode(context));
        props.put("isIPhoneX", 0);
        props.put("isIPhoneXMax", 0);
        props.put("isPad", tablet ? 1 : 0);
        props.put("isNotchScreen", cutout);
        props.put("isLowPowerMode", power != null && power.isPowerSaveMode() ? 1 : 0);
        props.put("orientation", orientation);
        props.put("screenOrientation", orientation);
        props.put("theme", theme);
        int separator = locale.indexOf('-');
        props.put("appLanguage", separator > 0 ? locale.substring(0, separator) : locale);
        props.put("appLocale", locale);
        props.put("isAppBackground", appInBackground);
        props.put("queryItems", Collections.emptyMap());
        props.put("statusBarHeight", insets.get("top"));
        props.put("navigationBarHeight", insets.get("bottom"));
        props.put("safeAreaHeight", insets.get("top"));
        return Collections.unmodifiableMap(props);
    }

    private static int accessibilityMode(@NonNull Context context) {
        AccessibilityManager manager = (AccessibilityManager) context.getSystemService(Context.ACCESSIBILITY_SERVICE);
        return manager != null && manager.isTouchExplorationEnabled() ? 1 : 0;
    }

    @NonNull
    private static Map<String, Double> systemInsets(@NonNull View host, float density) {
        WindowInsets current = host.getRootWindowInsets();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && current != null) {
            android.graphics.Insets values = current.getInsets(WindowInsets.Type.systemBars());
            return insets(values.top / density, values.right / density, values.bottom / density, values.left / density);
        }
        int top = current == null ? 0 : current.getSystemWindowInsetTop();
        int right = current == null ? 0 : current.getSystemWindowInsetRight();
        int bottom = current == null ? 0 : current.getSystemWindowInsetBottom();
        int left = current == null ? 0 : current.getSystemWindowInsetLeft();
        return insets(top / density, right / density, bottom / density, left / density);
    }

    @NonNull
    private static Map<String, Double> insets(double top, double right, double bottom, double left) {
        Map<String, Double> values = new HashMap<>();
        values.put("top", Math.max(0d, top));
        values.put("right", Math.max(0d, right));
        values.put("bottom", Math.max(0d, bottom));
        values.put("left", Math.max(0d, left));
        return Collections.unmodifiableMap(values);
    }
}

File: packages/sdk-ios/Sources/LynxShipGlobalProps.swift
import Foundation
import UIKit

/** Builds the reserved host context injected into an iOS Lynx container. */
enum LynxShipGlobalProps {
  static func create(
    host: UIView,
    containerID: String,
    containerInitTime: String,
    appProps: [String: Any],
    appInBackground: Bool
  ) -> [String: Any] {
    let screen = host.window?.screen ?? UIScreen.main
    let screenSize = screen.bounds.size
    let insets = host.safeAreaInsets
    let width = max(0, host.bounds.width)
    let height = max(0, host.bounds.height)
    let contentWidth = max(0, width - insets.left - insets.right)
    let contentHeight = max(0, height - insets.top - insets.bottom)
    let orientation = orientation(for: host, width: width, height: height)
    let theme: String
    switch host.traitCollection.userInterfaceStyle {
    case .dark: theme = "dark"
    case .light: theme = "light"
    default: theme = "system"
    }
    let locale = Locale.current.identifier
    let language = Locale.current.languageCode ?? locale.split(separator: "_").first.map(String.init) ?? locale
    let isTablet = UIDevice.current.userInterfaceIdiom == .pad
    let isNotch = UIDevice.current.userInterfaceIdiom == .phone && insets.top > 20
    var props = appProps
    props["os"] = "ios"
    props["osVersion"] = UIDevice.current.systemVersion
    props["deviceModel"] = UIDevice.current.model
    props["containerID"] = containerID
    props["containerInitTime"] = containerInitTime
    props["screenWidth"] = screenSize.width
    props["screenHeight"] = screenSize.height
    props["contentWidth"] = contentWidth
    props["contentHeight"] = contentHeight
    props["safeAreaInsets"] = ["top": insets.top, "right": insets.right, "bottom": insets.bottom, "left": insets.left]
    props["pixelRatio"] = screen.scale
    props["accessibleMode"] = UIAccessibility.isVoiceOverRunning ? 1 : 0
    props["isIPhoneX"] = isNotch ? 1 : 0
    props["isIPhoneXMax"] = isNotch ? 1 : 0
    props["isPad"] = isTablet ? 1 : 0
    props["isNotchScreen"] = isNotch
    props["isLowPowerMode"] = ProcessInfo.processInfo.isLowPowerModeEnabled ? 1 : 0
    props["orientation"] = orientation
    props["screenOrientation"] = orientation
    props["theme"] = theme
    props["appLanguage"] = language
    props["appLocale"] = locale
    props["isAppBackground"] = appInBackground
    props["queryItems"] = [String: String]()
    props["topHeight"] = insets.top
    props["bottomHeight"] = insets.bottom
    props["safeAreaHeight"] = insets.top
    return props
  }

  private static func orientation(for host: UIView, width: CGFloat, height: CGFloat) -> String {
    switch host.window?.windowScene?.interfaceOrientation {
    case .landscapeLeft: return "landscape-left"
    case .landscapeRight: return "landscape-right"
    case .portraitUpsideDown: return "portrait-upside-down"
    case .portrait: return "portrait"
    default: return width > height ? "landscape" : "portrait"
    }
  }
}
